### PR TITLE
NurseConnect Registration Flows

### PR DESCRIPTION
**Pull request by NicolaasDieBaas**

### CODE REVIEW

**Comment by NicolaasDieBaas** [comment]
@gsvr please review

**Comment by gsvr** [comment]
@NicolaasDieBaas Have you self-reviewed this?

**Comment by NicolaasDieBaas** [comment]
@imsickofmaps Please review

File: go-app-ussd_nurse.js [hunk 1]
```diff
@@ -72,6 +82,82 @@ go.app = function() {
...
+
+        self.jembi_json_api_call = function(method, params, payload, endpoint, im) {
+            var http = new JsonApi(im, {
+                auth: {
+                    username: im.config.jembi.username,
+                    password: im.config.jembi.password
+                }
+            });
+            switch(method) {
+                case "post":
```
Comment (imsickofmaps) [inline go-app-ussd_nurse.js]: I think we can drop this now because we don't post to Jembi from Jsbox apps anymore

File: go-app-ussd_nurse.js [hunk 2]
```diff
@@ -72,6 +82,82 @@ go.app = function() {
...
+        // temporary - TODO: adapt IdentityStore class in seed-jsbox-utils (#15) to have optin function
+        self.optin_identity = function(identity) {
+            var http = new JsonApi(self.im, {});
+            http.defaults.headers.Authorization = ['Token ' + self.im.config.services.identity_store.token];
+
+            var endpoint = 'optin/';
+            var url = self.im.config.services.identity_store.url + endpoint;
+
+            return http.post(url, {data: identity})
+                //.then(this.log_service_call('POST', url, identity, null))
```
Comment (imsickofmaps) [inline go-app-ussd_nurse.js]: Use or delete?

File: go-app-ussd_nurse.js [hunk 3]
```diff
@@ -72,6 +82,82 @@ go.app = function() {
...
+                .then(function(response) {
+                    return response.data;
+                });
+        },
+
+    // REGISTRATION FINISHED SMS HANDLING
+
+        self.send_registration_thanks = function(msisdn) {
```
Comment (imsickofmaps) [inline go-app-ussd_nurse.js:140]: Should be posting to the message_sender directly

File: go-app-ussd_nurse.js [hunk 4]
```diff
@@ -134,6 +220,197 @@ go.app = function() {
...
+                    .then(function(identity) {
+                        self.im.user.set_answer("registrant", identity);
+                        return self.states.create('state_check_optout_reg');
+                    });
+                }
+            });
+        });
+
+        self.add('state_check_optout_reg', function(name) {
+            var registrant_msisdn = Object.keys(self.im.user.answers.registrant.details.addresses.msisdn)[0];
```
Comment (imsickofmaps) [inline go-app-ussd_nurse.js]: Not sure this is safe using `[0]` - if the user has more than one msisdn (they could), this will fail.

Comment (imsickofmaps) [inline go-app-ussd_nurse.js]: Have they not already given us the msisdn we're looking for explicitly?

File: go-app-ussd_nurse.js [hunk 5]
```diff
@@ -134,6 +220,197 @@ go.app = function() {
...
+            //          + Object.keys(self.im.user.answers.registrant.details.addresses.msisdn)[0];;
+            //     }
+            // }
+
+            // identity PATCH..?
+
+            return Q
+            .all([
+                self.send_registration_thanks(
+                    Object.keys(self.im.user.answers.registrant.details.addresses.msisdn)[0]
```
Comment (imsickofmaps) [inline go-app-ussd_nurse.js]: Same

**Comment by imsickofmaps** [comment]
@NicolaasDieBaas sorry, I commented on the compiled file 😢

**Comment by NicolaasDieBaas** [comment]
@imsickofmaps no problem on commenting on the compiled file ;)

File: go-app-ussd_nurse.js (hunk 1 above)
Comment (NicolaasDieBaas) [reply go-app-ussd_nurse.js]: @imsickofmaps so that means we won't validate clinic (facility) codes..?

File: go-app-ussd_nurse.js (hunk 3 above)
Comment (NicolaasDieBaas) [reply go-app-ussd_nurse.js:140]: @imsickofmaps thought we might want to... seed-jsbox-utils will also consequently have to be updated.  Currently MessageSender only saves inbound messages...

Comment (imsickofmaps) [reply go-app-ussd_nurse.js:140]: @NicolaasDieBaas ah. Yes, we need that creation of outbound message. I'm going to update the post format though to include `identity` ASAP. We need that adding to the utils. Open a ticket for it?

File: go-app-ussd_nurse.js (hunk 1 above)
Comment (imsickofmaps) [reply go-app-ussd_nurse.js]: @NicolaasDieBaas Those are GET requests - just the POST case I was saying (to make sure it's obvious there is no submissions directly)

**Comment by NicolaasDieBaas** [comment]
@imsickofmaps please do 2nd pass review.

**Comment by NicolaasDieBaas** [comment]
@imsickofmaps please review.  Not sure whether dropping the lazytext for outbound message should've been done..?  Might we run into translations issue in the future?

File: src/ussd_nurse.js [hunk 6]
```diff
@@ -69,20 +87,78 @@ go.app = function() {
...
+        },
+
+    // REGISTRATION FINISHED SMS HANDLING
+
+        self.send_registration_thanks = function(msisdn) {
+            return ms.
+            create_outbound_message(
+                self.im.user.answers.registrant.id,
+                msisdn,
+                "Welcome to NurseConnect. For more options or to " +
```
Comment (imsickofmaps) [inline src/ussd_nurse.js:143]: Yes, please lazytext these

File: src/ussd_nurse.js [hunk 7]
```diff
@@ -131,6 +207,196 @@ go.app = function() {
...
+                    return choice.value;
+                }
+            });
+        });
+
+        self.add('state_save_nursereg', function(name) {
+            // Save useful identity info
+            self.im.user.answers.registrant.details.nurseconnect.is_registered = "true";
+
+            // if (self.im.user.answers.operator.id !== self.im.user.answers.registrant.id) {
```
Comment (imsickofmaps) [inline src/ussd_nurse.js]: If you're going to leave big chunks of commented out code in (I'd rather you didn't) please state the issue it expected to be uncommented in.

**Comment by imsickofmaps** [comment]
@NicolaasDieBaas two minor things then 👍

File: src/ussd_nurse.js (hunk 6 above)
Comment (NicolaasDieBaas) [reply src/ussd_nurse.js:143]: to be addressed in https://github.com/praekeltfoundation/ndoh-jsbox/issues/38

**Pull request merged**